### PR TITLE
Fetch multiples path and multiples or single passwords from vault

### PR DESCRIPTION
**Pull request by Rondineli**
https://github.com/Rondineli/serverless-vault-plugin/issues/8

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -49,12 +49,57 @@ custom:
     keyId: ${env:KEY_KMS_ID}
 ```
 
+### Using single path secrets with multiples password
```
Comment (pedrotcaraujo) [inline README.md]: passwords

File: README.md [hunk 2]
```diff
@@ -49,12 +49,57 @@ custom:
...
 
-### Add your env var as a list
+Then add your env var as a list
 ```
 environment:
   - VAR_FOO
   - VAR_BAR
 ```
 
+It needs an structure at vault like: `vault write secret/path VAR_FOO=foo VAR_BAR=bar
```
Comment (pedrotcaraujo) [inline README.md]: It needs `a` structure at vault`,` like

File: README.md [hunk 3]
```diff
@@ -49,12 +49,57 @@ custom:
...
+
+
+### Using multiple path secrets with multiples or single password
```
Comment (pedrotcaraujo) [inline README.md]: paths

File: README.md [hunk 4]
```diff
@@ -49,12 +49,57 @@ custom:
...
+```
+
+Then add your env var as a list
+```
+environment:
+  - path_foo
+  - path_bar
+```
+
+It needs an structure at vault like: `vault write secret/path_foo ANY_FOO=foo && vault write secret/path_bar ANY_BAR=bar`
```
Comment (pedrotcaraujo) [inline README.md]: It needs `a` structure at vault`,` like

File: index.js [hunk 5]
```diff
@@ -92,56 +93,114 @@ class ServerlessPlugin {
...
+            headers: {
+              'X-Vault-Token': this.serverless.service.custom.vault.token,
+              'Content-Type':'application/json',
+            },
+            strictSSL: this.serverless.service.custom.vault.ssl_check || false
+          };
+
+          request.get(myOptions, (error, response = {}, body) => {
+
+            var arrayData = [];
```
Comment (pedrotcaraujo) [inline index.js]: no reason to use `var` here, use `const` instead

File: index.js [hunk 6]
```diff
@@ -92,56 +93,114 @@ class ServerlessPlugin {
...
+            if (!error && typeof response.statusCode !== 'undefined' && response.statusCode == 200){
+              var data = JSON.parse(body);
```
Comment (pedrotcaraujo) [inline index.js]: no reason to use `var` here, use `const` instead

File: index.js [hunk 7]
```diff
@@ -92,56 +93,114 @@ class ServerlessPlugin {
...
+                    secret,
+                    data.data[secret]
+                  )
+                );
+              }
+              Promise.all(arrayData).then((result) => {
+                this.serverless.service.provider.environment = {};
+   
+                for (var rst in result) {
+                  var key = result[rst]['key'].toString();
```
Comment (pedrotcaraujo) [inline index.js]: no reason to use `var` here, use `const` instead

File: index.js [hunk 8]
```diff
@@ -92,56 +93,114 @@ class ServerlessPlugin {
...
+                  var value = result[rst]['value'].CiphertextBlob.toString('base64');
```
Comment (pedrotcaraujo) [inline index.js]: no reason to use `va`r here, use `const` instead

File: index.js [hunk 9]
```diff
@@ -92,56 +93,114 @@ class ServerlessPlugin {
...
+                );
+              } else  {
+                this.serverless.cli.log('Key ' + key + 'var not found on vault to be encrypted by kms');
+              }
             }
+            Promise.all(arrayData).then((result) => {
+              this.serverless.service.provider.environment = {};
+
+              for (var rst in result) {
+                var key = result[rst]['key'].toString();
```
Comment (pedrotcaraujo) [inline index.js]: no reason to use `var` here, use `const` instead

File: index.js [hunk 10]
```diff
@@ -92,56 +93,114 @@ class ServerlessPlugin {
...
+                var value = result[rst]['value'].CiphertextBlob.toString('base64');
```
Comment (pedrotcaraujo) [inline index.js]: no reason to use `var` here, use `const` instead

File: index.js [hunk 11]
```diff
@@ -92,56 +93,114 @@ class ServerlessPlugin {
...
+      });
+    } else {
+
+      return new Promise((resolve, reject) => {
+
+        var keysToVault = this.serverless.service.provider.environment;
+
+        this.serverless.service.provider.environment = [];
+
+        for (var key in keysToVault) {
```
Comment (pedrotcaraujo) [inline index.js]: avoid using `var` here, if you need to mutate a variable, use `let` instead

File: index.js [hunk 12]
```diff
@@ -92,56 +93,114 @@ class ServerlessPlugin {
...
+          };
+
+          request.get(myOptions, (error, response = {}, body) => {
+
+            var arrayData = [];
+            if (!error && typeof response.statusCode !== 'undefined' && response.statusCode == 200){
+              var data = JSON.parse(body);
+      
+
+              for (var secret in data.data) {
```
Comment (pedrotcaraujo) [inline index.js]: avoid using `var` here, if you need to mutate a variable, use `let` instead

File: index.js [hunk 13]
```diff
@@ -92,56 +93,114 @@ class ServerlessPlugin {
...
+                  this.kmsEncryptVariable(
+                    secret,
+                    data.data[secret]
+                  )
+                );
+              }
+              Promise.all(arrayData).then((result) => {
+                this.serverless.service.provider.environment = {};
+   
+                for (var rst in result) {
```
Comment (pedrotcaraujo) [inline index.js]: avoid using `var` here, if you need to mutate a variable, use `let` instead

File: index.js [hunk 14]
```diff
@@ -92,56 +93,114 @@ class ServerlessPlugin {
...
-            } else  {
-              this.serverless.cli.log('Key ' + key + 'var not found on vault to be encrypted by kms');
+      const myOptions = {
+        url: this.serverless.service.custom.vault.url + '/v1/' + this.serverless.service.custom.vault.secret,
+        method: 'GET',
+        headers: {
+          'X-Vault-Token': this.serverless.service.custom.vault.token,
+          'Content-Type':'application/json',
+        },
+        strictSSL: this.serverless.service.custom.vault.ssl_check || false
```
Comment (pedrotcaraujo) [inline index.js]: if `strictSSL` is expecting a `boolean` value. Use this code instead:
```javascript
strictSSL: !!this.serverless.service.custom.vault.ssl_check
```

File: index.js [hunk 15]
```diff
@@ -92,56 +93,114 @@ class ServerlessPlugin {
...
+      return new Promise((resolve, reject) => {
+        request.get(myOptions, (error, response = {}, body) => {
+          var arrayData = [];
+          if (!error && typeof response.statusCode !== 'undefined' && response.statusCode == 200){
+            var data = JSON.parse(body);
+            var keysToVault = this.serverless.service.provider.environment;
+
+            this.serverless.service.provider.environment = [];
+    
+            for (var key in keysToVault) {
```
Comment (pedrotcaraujo) [inline index.js]: avoid using `var` here, if you need to mutate a variable, use `let` instead

File: index.js [hunk 16]
```diff
@@ -92,56 +93,114 @@ class ServerlessPlugin {
...
-              var value = result[rst]['value'].CiphertextBlob.toString('base64');
-
-              this.serverless.service.provider.environment[key] = value;
+        });
+      });
+    } else {
+
+      return new Promise((resolve, reject) => {
+
+        var keysToVault = this.serverless.service.provider.environment;
```
Comment (pedrotcaraujo) [inline index.js]: no reason to use `var` here, use `const` instead

File: index.js [hunk 17]
```diff
@@ -92,56 +93,114 @@ class ServerlessPlugin {
...
+          'Content-Type':'application/json',
+        },
+        strictSSL: this.serverless.service.custom.vault.ssl_check || false
+      };
+
+      return new Promise((resolve, reject) => {
+        request.get(myOptions, (error, response = {}, body) => {
+          var arrayData = [];
+          if (!error && typeof response.statusCode !== 'undefined' && response.statusCode == 200){
+            var data = JSON.parse(body);
```
Comment (pedrotcaraujo) [inline index.js]: no reason to use `var` here, use `const` instead

File: index.js [hunk 18]
```diff
@@ -92,56 +93,114 @@ class ServerlessPlugin {
...
+            var keysToVault = this.serverless.service.provider.environment;
```
Comment (pedrotcaraujo) [inline index.js]: no reason to use `var` here, use `const` instead

File: index.js [hunk 19]
```diff
@@ -92,56 +93,114 @@ class ServerlessPlugin {
...
+        headers: {
+          'X-Vault-Token': this.serverless.service.custom.vault.token,
+          'Content-Type':'application/json',
+        },
+        strictSSL: this.serverless.service.custom.vault.ssl_check || false
+      };
+
+      return new Promise((resolve, reject) => {
+        request.get(myOptions, (error, response = {}, body) => {
+          var arrayData = [];
```
Comment (pedrotcaraujo) [inline index.js]: no reason to use `var` here, use `const` instead

File: README.md (hunk 1 above)
Comment (Rondineli) [reply README.md]: Done

**Comment by Rondineli** [comment]
@pedrotcaraujo :+1: Done!

File: index.js [hunk 20]
```diff
@@ -87,7 +88,7 @@ class ServerlessPlugin {
    * @private
    * @returns {Promise}
    */
-  requestData(token = null) {
+  getSingleSecrets(token = null ) {
```
Comment (pedrotcaraujo) [inline index.js]: default assignment is not needed, once you don't pass a parameter to the function, the parameter will be `undefined` by default

File: index.js [hunk 21]
```diff
@@ -144,21 +145,93 @@ class ServerlessPlugin {
     });
   }
 
+  /**
+   * @private
+   * @returns {Promise}
+   */
+  getMultiplesSecrets(token = null ) {
```
Comment (pedrotcaraujo) [inline index.js]: default assignment is not needed, once you don't pass a parameter to the function, the parameter will be `undefined` by default

**Review by pedrotcaraujo — APPROVED** [review]
👍

**Pull request merged**